cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -329,985 +302,602 @@ struct ValueConverter<DurationType> {
+    return AppendNdarrayTyped<TYPE, NUMPY_TYPE>(ndarray);       \
   }
 
 // Use internal::VisitSequence, fast for NPY_OBJECT but slower otherwise
-#define LIST_SLOW_CASE(TYPE)                            \
-  case Type::TYPE: {                                    \
-    return value_converter_->Extend(obj, value_length); \
+#define LIST_SLOW_CASE(TYPE_ID)                               \
+  case Type::TYPE_ID: {                                       \
+    return Extend(this->value_converter_.get(), value, size); \
   }
 
-// Base class for ListConverter and FixedSizeListConverter (to have both work with CRTP)
-template <typename TypeClass, NullCoding null_coding>
-class BaseListConverter : public TypedConverter<TypeClass, null_coding> {
+template <typename T>
+class PyListConverter : public ListConverter<T, PyConverter, PyConverterTrait> {
  public:
-  using BuilderType = typename TypeTraits<TypeClass>::BuilderType;
-
-  explicit BaseListConverter(bool from_pandas, bool strict_conversions,
-                             bool ignore_timezone)
-      : from_pandas_(from_pandas),
-        strict_conversions_(strict_conversions),
-        ignore_timezone_(ignore_timezone) {}
-
-  Status Init(ArrayBuilder* builder) override {
-    this->builder_ = builder;
-    this->typed_builder_ = checked_cast<BuilderType*>(builder);
-
-    this->value_type_ = checked_cast<const TypeClass&>(*builder->type()).value_type();
-    RETURN_NOT_OK(GetConverter(value_type_, from_pandas_, strict_conversions_,
-                               ignore_timezone_, &value_converter_));
-    return this->value_converter_->Init(this->typed_builder_->value_builder());
-  }
+  Status Append(PyObject* value) override {
+    if (PyValue::IsNull(this->options_, value)) {
+      return this->list_builder_->AppendNull();
+    }
 
-  template <int NUMPY_TYPE, typename Type>
-  Status AppendNdarrayTypedItem(PyArrayObject* arr) {
-    using traits = internal::npy_traits<NUMPY_TYPE>;
-    using T = typename traits::value_type;
-    using ValueBuilderType = typename TypeTraits<Type>::BuilderType;
+    RETURN_NOT_OK(this->list_builder_->Append());
+    if (PyArray_Check(value)) {
+      RETURN_NOT_OK(AppendNdarray(value));
+    } else if (PySequence_Check(value)) {
+      RETURN_NOT_OK(AppendSequence(value));
+    } else {
+      return internal::InvalidType(
+          value, "was not a sequence or recognized null for conversion to list type");
+    }
 
-    const bool null_sentinels_possible =
-        // Always treat Numpy's NaT as null
-        NUMPY_TYPE == NPY_DATETIME || NUMPY_TYPE == NPY_TIMEDELTA ||
-        // Observing pandas's null sentinels
-        (from_pandas_ && traits::supports_nulls);
+    return ValidateBuilder(this->list_type_);
+  }
 
-    auto child_builder = checked_cast<ValueBuilderType*>(value_converter_->builder());
+ protected:
+  Status ValidateOverflow(const MapType*, int64_t size) { return Status::OK(); }
```

Comment:
Since MapArray has the same offset structure as ListArray it should probably have non empty overflow validation

Reply:
Added a test case for it but takes 20 minutes to run.